modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java

```diff
@@ -99,11 +99,14 @@ public IngestDocument execute(IngestDocument ingestDocument) {
 
         ZonedDateTime dateTime = null;
         Exception lastException = null;
-        for (Function<Map<String, Object>, Function<String, ZonedDateTime>> dateParser : dateParsers) {
+        int dateParserIndex = 0;
+        while (dateTime == null && dateParserIndex < dateParsers.size()) {
+            var dateParser = dateParsers.get(dateParserIndex);
```

Comment:
I'm not sure I understand why we do it this way, why doesn't it work to do:

```java
for (Function<stuff> dateParser : dateParsers) {
    try {
        dateTime = dateParser.apply(ingestDocument.getSourceAndMetadata()).apply(value);
        break;
    catch (Exception e) {
        ...
    }
}
```

To just short circuit for the first non-error-throwing date parser?

Reply:
Anything that terminates the loop after the parser succeeds will work.

Reply:
Okay, that seems more readable to me than a counter than increments and the null check, what do you think?

Reply:
I like having the loop termination criteria explicitly specified upfront with a while loop but I don't feel that strongly about it.